tiledb/sparse_array.py

```diff
@@ -0,0 +1,455 @@
+import warnings
+
+import numpy as np
+
+import tiledb
+import tiledb.cc as lt
+
+from .array import (
+    check_for_floats,
+    index_as_tuple,
+    index_domain_subarray,
+    replace_ellipsis,
+    replace_scalars_slice,
+)
+from .libtiledb import Array
+
+
+class SparseArrayImpl(Array):
+    """Class representing a sparse TileDB array (internal).
+
+    Inherits properties and methods of :py:class:`tiledb.Array`.
+    """
+
+    def __init__(self, *args, **kw):
+        super().__init__(*args, **kw)
+        if not self.schema.sparse:
+            raise ValueError("Array at '{}' is not a sparse array".format(self.uri))
+
+    def __len__(self):
+        raise TypeError("SparseArray length is ambiguous; use shape[0]")
+
+    def __setitem__(self, selection, val):
+        """Set / update sparse data cells
+
+        :param tuple selection: N coordinate value arrays (dim0, dim1, ...) where N in the ndim of the SparseArray,
+            The format follows numpy sparse (point) indexing semantics.
+        :param value: a dictionary of nonempty array attribute values, values must able to be converted to 1-d numpy arrays.\
+            if the number of attributes is one, then a 1-d numpy array is accepted.
+        :type value: dict or :py:class:`numpy.ndarray`
```

Comment:
This mismatch between the arg name and doc existed prior to this PR, but let's correct it here

```suggestion
    def __setitem__(self, selection, val):
        """Set / update sparse data cells

        :param tuple selection: N coordinate value arrays (dim0, dim1, ...) where N in the ndim of the SparseArray,
            The format follows numpy sparse (point) indexing semantics.
        :param val: a dictionary of nonempty array attribute values, values must able to be converted to 1-d numpy arrays.\
            if the number of attributes is one, then a 1-d numpy array is accepted.
        :type val: dict or :py:class:`numpy.ndarray`
```